### DISCUSSION TITLE
Expand env variables in clink settings file

### DISCUSSION THREAD

**Discussion by 3N4N** [chrisant996/clink#322] [Q&A] [answered]
The clink settings file -- is there a way to expand env variables in it? The following doesn't work.

```
clink.path = %LOCALAPPDATA%\clink\clink-completions
```

If it is not supported, could you count it as a feature request? It'd be useful for those who version-controls their config files. User-agnostic config is helpful.

**Comment by chrisant996** [comment, marked as answer]
Have you considered `~\appdata\local\clink...etc`?

**Reply by 3N4N** [reply]
Isn't working. I tried both forward and backward slashes. Tried quoting and not quoting. The completions don't work with `~\AppData\Local\clink\...`.

**Reply by 3N4N** [reply]
It works now, in v1.3.45.83a752.

**Comment by chrisant996** [comment]
I didn't mean to literally use "...". That was meant as "fill in the rest depending on what you're trying to do".

My point was that you're trying to specify a path under your user profile directory, so you could replace the environment variable portion of the path with "~\appdata\local".

**Reply by 3N4N** [reply]
I didn't literally use the ellipses. I don't think such laymen would use clink.

**Reply by chrisant996** [reply]
"Isn't working" isn't very specific. Can you share more detail about what you did, and what results occurred?

**Reply by chrisant996** [reply]
I was answering from my phone before. I had a chance to get to a computer and double check, and there is a bug: tilde is only expanded for the profile directory. I'm updating it so all script paths expand tilde.

**Reply by chrisant996** [reply]
Also, it turns out that Readline and Bash replace {space}{tilde} even in the middle of a path name, which results in a malformed path name.  I'm not going to change how Readline works, because I don't want to risk unexpected side effects inside Readline or inconsistent editing behavior (I can see edge cases where it could potentially be considered a quirk that can be exploited for benefit).

But I am rewriting how Clink handles tilde expansion outside of the Readline library.  Most tilde expansions will be affected, but things purely inside Readline such as the inputrc file path and <kbd>Alt</kbd>-<kbd>&</kbd> for example will still produce malformed paths from embedded {space}{tilde} sequences.

Anyway, thanks for reporting this and helping to uncover multiple problems in tilde expansion in various places in Clink and Readline.

**Comment by chrisant996** [comment]
> It'd be useful for those who version-controls their config files. User-agnostic config is helpful.

I version control my files. I go to extreme lengths with that. I've had my whole user config under source control for almost 30 years, and I nearly daily use source control to replicate config changes across 4 computers (and less frequently across VMs and USB memory sticks, etc).

In Clink, due to its Unix roots, tilde is a good way to accomplish user agnostic config.

Also, in whatever script where you set your other user-specific config environment variables, you could set CLINK_PATH.

**Reply by 3N4N** [reply]
Wow. Okay dude. Bye. Thank for your life story and educating me on how awesome you are.

**Reply by chrisant996** [reply]
> Wow. Okay dude. Bye. Thank for your life story and educating me on how awesome you are.

I'm sorry that's how it came across. I'm trying to help. I was trying to convey that there is a way to do this without needing to introduce complexity of environment variable evaluation into Clink setting strings (there are a bunch of side issues and edge cases that would be introduced), and tilde satisfies the requirement of user agnostic configuration shared between multiple computers via source control.

A solution already exists. Is the existing solution sufficient?

If so, can it be used?

If not, can more information be shared about what has been tried, and specifically what isn't working about it?

For the record, given how many "Fixed this" and "Fixed that" items I have to write in the change logs for Clink, do you really think I'm trying to say I'm awesome...? 🤔😂

**Comment by 3N4N** [comment]
@chrisant996 I forgot the whole thing about git completions. Updated both clink and clink-completions today and it works with `clink.path` works with tilde expansion. It still doesn't with cmd-style `%var%` expansion (I don't need it; just wanted to document it here for other viewers).

Thanks for the updates.
